### ISSUE TITLE
Shell: navigation model

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#9035]
## Overview

This is a proposal for a complete navigation model in dagger shell. It defines:

- Filesystem navigation
- Module loading
- Command execution

And the relationship between those features.

## Filesystem navigation

- The dagger shell presents a virtual filesystem to the user, which they can navigate
- Navigation is done by changing the shell's current location in the virtual filesystem: the working directory or *workdir*.
- A location in virtual filesystem has 2 components: *context* and *path*.
- A context is a virtual filesystem which acts as a sandbox for convenience and safety. The most common kinds of contexts are git repositories - either local or remote.

### Initial context

- When the shell is started, it gets an initial context
- By default, the initial context is inferred from the working directory in the system, by finding an enclosing git repository or dagger module. The initial context is *not* the entire host filesystem
- Alternatively, the shell can be started with a custom workdir (`dagger shell --workdir PATH|URL`) or with no workdir (`--no-workdir`)

### Navigating within a context

- `.cd PATH` changes the workdir within the current context.
- `.cd /` navigates to the root of the context, *not* of the host filesystem.
- The context root cannot be escaped
- `.`,  `..`  in a path have the usual meaning, eg. `./foo`, `../../foo` etc.

### Navigating across contexts

Dagger can navigate to other contexts beyond the initial context.

- `.cd URL` changes the current context and workdir to a remote filesystem. For example `.cd github.com/dagger/dagger/docs@v0.13.0` changes the context to `github.com/dagger/dagger@v0.13.0` and the workdir to `/docs`.
- `.cd` with no argument navigates back to the initial context

### Other navigation builtins

- `.back [N]` goes back N steps in the navigation (default 1). Like the back button in a browser :)
- `.ls [PATH|URL]` lists contents of the specified directory
- `.pwd` prints the current workdir
- The workdir is shown in the prompt

## Module loading

Modules are automatically loaded from the workdir. For example, navigating to the dagger repo will automatically load the corresponding module:

```console
⋈ .cd github.com/dagger/dagger@v0.13.0
github.com/dagger/dagger@v0.13.0 ⋈ 
```

## Command execution

To execute each command in a pipeline, the dagger shell looks up the corresponding function from the current context, and calls it. Function lookup follows this sequence:

1. If the command is the name of a function in the current object, the function is executed

2. If the command is the name of a dependency in the current module, the dependency is loaded and its constructor executed

3. If the command is the name of a function in the *stdlib*, the function is executed

4. If the command is a local path within a module, that module is loaded its constructor executed. Relative paths are interpreted from the shell's working directory, like any other tool. Note: `.` is a valid path.

5. If the command is the address of a remote module, that module is loaded and its constructor executed. For example: `⋈ github.com/shykes/daggerverse/wolfi | container | with-file /bin/dagger $( github.com/dagger/dagger@v0.13.0 | cli | binary)`

### Command disambiguation

Since the same command could be available in different namespaces (current object; dependencies; stdlib; local path) sometimes it's necessary to disambiguate.

The following builtins are available for this purpose. They each return a special object, which corresponds to one of the namespaces.

1. Current object only: no builtin for that.

2. `.deps` returns the current module's dependencies, as stored in the module's `dagger.json`. Example: `.deps | go | build`

3. `.stdlib` returns the stdlib. The stdlib is a facade concept that allows the shell to expose the parts of the core API we like, hide others, and possibly add new calls. It could also include hardcoded references to external modules, as we discussed doing in the past. In other words, it solves a short-term problem, while creating an elegant interface to a long-term solution we wanted to build anyway. Example: `.stdlib | go | build`

## Stdlib

`.stdlib` is the successor of `.core`. It reduces the role of builtins to exposing CLI functionality, and *not* mapping to core calls. As a result, the number of builtins is greatly diminished.

Stdlib reference:

```
⋈ .stdlib | .doc
container     container subsystem
git           git subsystem
http          http subsystem
host          access host system
...
```

Example pipeline using the stdlib:

```
⋈ container | from alpine | with-exec ls,-l | stdout
```

The same command with disambiguation:

```
⋈ .stdlib | container | from alpine | with-exec ls,-l | stdout
```

**Comment by vito** [comment]
Quick suggestion: `.cd -` should be same as `.back 1`, and repeating it should bring you back to where you were. (This is a common shell built-in behavior.)

**Comment by helderco** [comment]
If you need to set the constructor for the current module, you'll need to use `.` right? Implied without arguments if ommitted?

**Comment by shykes** [comment]
> If you need to set the constructor for the current module, you'll need to use `.` right? Implied without arguments if ommitted?

Yes correct. This could be complemented by #6323 , I know that is being actively discussed, cc @marcosnils @jedevc

**Comment by helderco** [comment]
> `⋈ .stdlib | .doc`

What should happen if you run `.stdlib` or `.deps` without arguments? Redundant with `.stdlib | .help`, or one throws an error pointing to use the other?

> `.stdlib is the successor of .core`

What about those not listed in `.stdlib`? Just hidden but accessible, or not even accessible? Can we have a way to list and access all core functions? Some of those are useful for different use cases, even if it's just an SDK author trying to figure out the API. That's why `.help` shows the most important ones today, but `.core` shows all.

**Comment by shykes** [comment]
> > `⋈ .stdlib | .doc`
> 
> What should happen if you run `.stdlib` or `.deps` without arguments? Redundant with `.stdlib | .help`, or one throws an error pointing to use the other?

It should be the same result whenever you call a function that returns an object. At the moment it prints the type & fields.. If we changed that behavior to showing the help (definitely an option) then it would apply consistently to `.stdlib` and every other.

> > `.stdlib is the successor of .core`
> 
> What about those not listed in `.stdlib`? Just hidden but accessible, or not even accessible? Can we have a way to list and access all core functions? Some of those are useful for different use cases, even if it's just an SDK author trying to figure out the API. That's why `.help` shows the most important ones today, but `.core` shows all.

Yeah I think that makes sense. Maybe we rename it to `sys` to better map to the stdlib/system calls analogy?

**Comment by helderco** [comment]
> It should be the same result whenever you call a function that returns an object. At the moment it prints the type & fields.. If we changed that behavior to showing the help (definitely an option) then it would apply consistently to `.stdlib` and every other.

Yeah, I was thinking it would just show "Available Functions", while `.stdlib | .help` maybe shows more information than that. If we no longer print object fields with https://github.com/dagger/dagger/issues/8990, could be the same with any object.

**Comment by helderco** [comment]
We need to decide what to do with the `--workdir` flag since it conflicts with a global flag. The global flag is hidden. Maybe we could remove it from there?

**Comment by helderco** [comment]
Still missing filesystem navigation.

**Comment by helderco** [comment]
The filesystem navigation is in working condition in:
- https://github.com/dagger/dagger/pull/9511

@shykes, can you please try it to validate the design?

There may still be some bugs but I don’t want to waste time investing in adding tests that may be discarded anyway due to possible design changes. See how it “feels” first.

I think that using `.cd`, `.ls`, and `.pwd` can feel a bit awkward because of the extra `.` prefix, but may also feel limiting because you may feel the desire to do more like you’re used to in your own shell, like read a file’s contents, fuzzy search files, etc…

However, I do love how much simpler it becomes when referring to paths, like loading a module, calling the constructor, or providing a `Directory` argument. The reason for that is we can make those arguments much shorter if your workdir is near.

**Comment by helderco** [comment]
There's one other thing that may trip you up.

Let's say I'm working with the `sdk/python/dev` module. I have a function for linting, and by default it lints the `sdk/python` directory (i.e., `+defaultPath=".."`). 

But now I want to navigate to `/docs` and look for snippets to lint, like:
```
(python-sdk-dev) sdk/python/dev ⋈ .cd /docs/current_docs
(dagger-dev) docs/current_docs ⋈ lint script.py
```
That won't work because as soon as I navigate out of `sdk/php/dev`, it looks for a module in the parent directories and loads it (the root `dagger-dev` module in this case).

**Comment by helderco** [comment]
If you feel limited by the builtin commands for navigation, you can mount the current folder in a container and open a terminal, like so:
```
(php-sdk) sdk/php ⋈ directory | with-directory / . | terminal
● Attaching terminal:
    directory: Directory!
    Directory.withDirectory(
    │ │ directory: Host.directory(path: "/Users/helder/Projects/dagger/sdk/php"): Directory!
    │ │ path: "/"
    │ ): Directory!

dagger /src $ ls
CHANGELOG.md        composer.json       dev                 generated           psalm.xml           src
LICENSE             composer.lock       docker              phpcs.xml           runtime             tests
README.md           dagger.json         docker-compose.yml  phpunit.xml.dist    scripts             vendor
dagger /src $
```

Maybe we could provide a shorthand convenience for `directory | with-directory / .`:
```
(php-sdk) sdk/php ⋈ .cwd | glob 'src/**/*.php'
...
(php-sdk) sdk/php ⋈ .cwd | terminal
dagger /src $ cat dagger.json
...
```

Or just `.terminal [path]`.
